### ISSUE TITLE
Allow intersections of `callable` and `object`

### ISSUE THREAD

**Issue by MidnightDesign** [vimeo/psalm#6911]
Consider the following function:

```php
/**
 * @param list<(callable(object): void)&object> $handlers
 */
function buildHandlerMap(array $handlers): array {
    $handlersByClassName = [];
    foreach ($handlers as $handler) {
        $handlersByClassName[get_class($handler)] = $handler;
    }
    return $handlersByClassName;
}
```
Playground snippet: https://psalm.dev/r/605d7778e7

*Expected:* Psalm doesn't complain because objects can be callables (`__invoke`)

*Actual:* Psalm reports "Intersection types must be all objects, Psalm\Type\Atomic\TCallable provided in docblock for buildHandlerMap"

It also doesn't help to just use `callable` because Psalm then (rightfully) complains about not passing an object to `get_class()`.

**Comment by orklah** [comment]
Psalm has a `callable-object` type for that.

Unfortunately: https://psalm.dev/r/6a61786509 it doesn't work in this context. Apparently, it's not expected here: https://github.com/vimeo/psalm/blob/b29737f2b154748430250effdbcf47de8a5904df/src/Psalm/Internal/Type/ParseTreeCreator.php#L749

I guess adding it to the list would solve the issue. Care for a PR?

**Comment by MidnightDesign** [comment]
> Psalm has a `callable-object` type for that.

Thanks, didn't know that. [It doesn't seem do be documented either.](https://www.google.com/search?q=site%3Apsalm.dev+%22callable-object%22)

Still, I think Psalm should accept `callable&object`. Maybe it could normalize it to `callable-object` internally. Also, what if I don't want to just use `object` but something more complex like `callable&MyFoo<string>`?

**Comment by orklah** [comment]
> It doesn't seem do be documented either.

Well, it doesn't currently work so that's coherent :) It seems to be an internal representation more than a type that is meant to be used like in your case. If you create a PR, you can create the documentation as well

> Still, I think Psalm should accept callable&object

Unless some motivated contributor chime in, it will probably not happen. As the error says, intersections types are currently limited to object (or all arrays). Intersections between pseudo types is not structurally possible in the type system right now.

> what if I don't want to just use object but something more complex like callable&MyFooInterface<string>?

If your interface declare an __invoke method, you don't need the callable intersection, if it doesn't, that's a really weird use case and I'd suggest creating an interface specifically for your callable classes

**Comment by MidnightDesign** [comment]
Thank you for taking the time to reply!

I don't know if it makes sense to leave this open as a feature request or if you want to close it since it doesn't seem likely anyone would want to implement it.

**Comment by orklah** [comment]
Well, if you don't intend to propose a PR to add `callable-object`, I think I'll close this
